File: day6/6.py
import string
from timeit import default_timer as timer

def get_input():
    with open("./day6/6.in","r") as f:
        return f.read().split("\n\n")

def p1():
    res = 0
    data = [group.replace("\n","") for group in get_input()]
    for group in data:
        res += len(set(group))
    return res

def p2():
    res = 0
    ques = set(string.ascii_lowercase)
    data  = [group.split("\n")for group in get_input()]
    for group in data:
        for line in group:
            ques &= set(line) # removes questions that werent answered yes. 
        res += len(ques)
        ques = set(string.ascii_lowercase)
    return res


print("Day 6: Custom Customs")

p1start = timer()
p1 = p1()
p1end = timer()
print(f"Part 1: {p1} in {p1end-p1start}s.")

p2start = timer()
p2 = p2()
p2end = timer()
print(f"Part 2: {p2} in {p2end-p2start}s.")

File: 2020/day6/6.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"time"

	"../helpers"
)

func getInput(fileName string) ([]string, error) {
	// Split by 2 line so can't use standard method
	data, err := ioutil.ReadFile(fileName)
	if err != nil {
		return nil, err
	}

	lines := strings.Split(string(data), "\n\n")
	var contents []string
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		contents = append(contents, string(line))
	}
	return contents, nil
}

func removeDuplicates(input string) []string {
	var found []string
	for _, char := range input {
		if !(helpers.Find(found, string(char))) {
			found = append(found, string(char))
		}
	}
	return found
}

func getDuplicates(a, b string) string {
	var found string
	for _, char := range a {
		if strings.Contains(b, string(char)) {
			found = found + string(char)
		}
	}
	return found
}

func p1(data []string) {
	defer helpers.TimeTrack(time.Now())
	var res int
	for _, group := range data {
		contains := removeDuplicates(strings.ReplaceAll(group, "\n", ""))
		res = res + len(contains)
	}
	fmt.Printf("Part 1: %d ", res)
}

func p2(data []string) {
	defer helpers.TimeTrack(time.Now())
	var res int
	var splitData [][]string
	for _, v := range data {
		splitData = append(splitData, strings.Split(v, "\n"))
	}
	for _, group := range splitData {
		ques := "abcdefghijklmnopqrstuvwxyz"
		for _, line := range group {
			ques = getDuplicates(ques, line)
		}
		res = res + len(ques)

	}
	fmt.Printf("Part 2: %d ", res)
}

func main() {
	fmt.Println("Day 6: Custom Customs")
	fileName := "/Users/jack/Documents/Github/aoc-2020/day6/6.in"
	data, err := getInput(fileName)
	if err != nil {
		panic(err)
	}
	p1(data)
	p2(data)
}
